### PR TITLE
[Audit policy] Support Event Hub and Log Analytics in server and database audit policy commands

### PR DESCRIPTION
**Pull request by Oleza1972**
**Description**

In this PR the following commands have been extended to support Event Hub and Log Analytics:

- az sql server audit-policy show:
- az sql db audit-policy show:
An extended JSON body is returned now. 
See: 
https://docs.microsoft.com/en-us/rest/api/sql/server%20auditing%20settings/get
https://docs.microsoft.com/en-us/rest/api/sql/database%20auditing%20settings/get

- az sql server audit-policy update,
- az sql db audit-policy update:

A new command line arguments have been added:
--blob-storage-target-state - enables\disables blob storage as auditing logs target
--log-analytics-target-state - enables\disables log analytics as auditing logs target
--log-analytics-workspace-resource-id - log analytics workspace resource id
--event-hub-target-state - enables\disables event hub as auditing logs target
--event-hub-authorization-rule-id - event hub authorization rule id
--event-hub - event hub name

Examples of commands:

sql db audit-policy update -g my_resource_group -s my_server -n my_database --state Enabled --blob-storage-target-state Enabled --storage-endpoint https://mystorageaccount.blob.core.windows.net/

sql db audit-policy update -g my_resource_group -s my_server -n my_database --state Enabled --blob-storage-target-state Disabled

sql server audit-policy update -g my_resource_group -n my_server --state Enabled --log-analytics-target-state Enabled --log-analytics-workspace-resource-id /subscriptions/1111-2222-3333-444-5555555/resourcegroups/my_resource_group/providers/microsoft.operationalinsights/workspaces/my_workspace

sql server audit-policy update -g my_resource_group -n my_server --state Enabled --log-analytics-target-state Disabled --event-hub-target-state Enabled --event-hub-authorization-rule-id /subscriptions/1111-2222-3333-4444/resourceGroups/my_resource_group/providers/Microsoft.EventHub/namespaces/my_eventhub_namespace/authorizationrules/my_eventhub_auth_rule_name --event-hub my_event_hub_name

**Testing Guide**

Unit tests test_sql_server_security_mgmt and test_sql_db_security_mgmt cover the test for the change.

**History Notes**

[SQL] az sql db audit-policy show: extend to show database's audit policy including LA and EH data
[SQL] az sql db audit-policy update: extend to allow LA and EH update along with database's audit policy
[SQL] az sql db audit-policy wait: place the CLI in a waiting state until a condition of the database's audit policy is met.
[SQL] az sql server audit-policy show: extend to show servers's audit policy including LA and EH data
[SQL] az sql server audit-policy update: extend to allow LA and EH update along with server's audit policy
[SQL] az sql server audit-policy wait: place the CLI in a waiting state until a condition of the server's audit policy is met.

### CODE REVIEW

**Comment by yonzhan** [comment]
sql

File: src/azure-cli/azure/cli/command_modules/sql/_params.py [hunk 1]
```diff
@@ -1328,6 +1355,33 @@ def _configure_security_policy_storage_params(arg_ctx):
...
+        c.argument('event_hub_target_state',
+                   arg_group=policy_arg_group,
+                   help='Indicates whether event hub is a destination for audit records.',
+                   arg_type=get_enum_type(BlobAuditingPolicyState))
+
+        c.argument('event_hub_authorization_rule_id',
+                   arg_group=policy_arg_group,
+                   help='The resource Id for the event hub authorization rule.')
+
+        c.argument('event_hub_name',
```
Comment (jaredmoo) [inline src/azure-cli/azure/cli/command_modules/sql/_params.py:1381]: Arguments should not have `name` in their name. E.g. we have args like `--server` or `--database` as opposed to `--server-name` or `--database-name`

Comment (Oleza1972) [reply src/azure-cli/azure/cli/command_modules/sql/_params.py]: fixed

File: src/azure-cli/azure/cli/command_modules/sql/_params.py [hunk 2]
```diff
@@ -257,6 +261,47 @@ def get_location_type_with_default_from_resource_group(cli_ctx):
     help='The number of readonly replicas to provision for the database. '
     'Only settable for Hyperscale edition.')
 
+blob_storage_target_state_param_type = CLIArgumentType(
+    arg_group=storage_arg_group,
+    options_list=['--blob-storage-target-state', '--bsts'],
+    configured_default='sql-server',
+    help='Indicates whether blob storage is a destination for audit records.',
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/_params.py:268]: Active voice. Inditicates -> Indicate

File: src/azure-cli/azure/cli/command_modules/sql/commands.py [hunk 3]
```diff
@@ -272,9 +271,8 @@ def load_command_table(self, _):
                             server_blob_auditing_policies_operations,
                             client_factory=get_sql_server_blob_auditing_policies_operations) as g:
 
-        g.show_command('show', 'get')
-        g.generic_update_command('update',
-                                 custom_func_name='server_audit_policy_update')
+        g.custom_show_command('show', 'server_audit_policy_show')
+        g.custom_command('update', 'server_audit_policy_update', supports_no_wait=True)
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/commands.py:275]: please add wait command 
See wait_command in https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md#register-commands

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 4]
```diff
@@ -1583,98 +1585,796 @@ def _db_security_policy_update(
...
+        cmd,
+        resource_group_name,
+        server_name,
+        database_name=None):
+    '''
+    Common code to get server or database diagnostic settings
+    '''
+
+    diagnostic_settings_url = _get_diagnostic_settings_url(cmd, resource_group_name, server_name, database_name)
+    azure_monitor_client = get_mgmt_service_client(cmd.cli_ctx, MonitorManagementClient)
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py:1630]: Try to reuse the following function.

https://github.com/Azure/azure-cli/blob/12c1c4c7d049e85d7fcdc4e46b37e45d2f1dab09/src/azure-cli/azure/cli/command_modules/monitor/_client_factory.py#L8-L11

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 5]
```diff
@@ -1583,98 +1585,796 @@ def _db_security_policy_update(
...
+
+    # Request audit policy
+    if database_name is None:
+        audit_policy = client.get(
+            resource_group_name,
+            server_name)
+    else:
+        audit_policy = client.get(
+            resource_group_name,
+            server_name,
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py:1665]: prefer no positional arguments

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 6]
```diff
@@ -1583,98 +1585,796 @@ def _db_security_policy_update(
...
+def server_audit_policy_show(
+        cmd,
+        client,
+        server_name,
+        resource_group_name):
+    '''
+    Show server audit policy
+    '''
+
+    return _audit_policy_show(
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py:1715]: same as above

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 7]
```diff
@@ -1583,98 +1585,796 @@ def _db_security_policy_update(
...
+    '''
+    Show database audit policy
+    '''
+
+    return _audit_policy_show(
+        cmd,
+        client,
+        resource_group_name,
+        server_name,
+        database_name)
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py:1737]: same as above. no positional arguments

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 8]
```diff
@@ -1583,98 +1585,796 @@ def _db_security_policy_update(
...
+    else:
+        diagnostic_settings = None
+        rollback_data = None
 
-    # Set other properties
-    if audit_actions_and_groups:
-        instance.audit_actions_and_groups = audit_actions_and_groups
+    # Update auditing global settings
+    try:
+        return _audit_policy_update_global_settings(
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py:2262]: no positional arguments

**Review by Juliehzl — CHANGES_REQUESTED** [review]
please resolve all my comments and change the title format.
And edit the history notes part in description, e.g.
```
History Notes

[Compute] sig image-definition create: add --features
[Compute] New API version of gallery_images 2020-09-30
```
in https://github.com/Azure/azure-cli/pull/15549#issue-504496171

File: src/azure-cli/azure/cli/command_modules/sql/_params.py (hunk 2 above)
Comment (Oleza1972) [reply src/azure-cli/azure/cli/command_modules/sql/_params.py:268]: Done

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 9]
```diff
@@ -22,29 +22,56 @@
 short-summary: Manage a database's auditing policy.
 """
 
+helps['sql db audit-policy wait'] = """
+type: command
+short-summary: Place the CLI in a waiting state until a condition of the database's audit policy is met.
+"""
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/_help.py:28]: Could you add an example here?

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 10]
```diff
@@ -980,29 +1007,56 @@
 short-summary: Manage a server's auditing policy.
 """
 
+helps['sql server audit-policy wait'] = """
+type: command
+short-summary: Place the CLI in a waiting state until a condition of the server's audit policy is met.
+"""
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/_help.py:1013]: same

File: src/azure-cli/azure/cli/command_modules/sql/commands.py [hunk 11]
```diff
@@ -260,9 +260,9 @@ def load_command_table(self, _):
                             database_blob_auditing_policies_operations,
                             client_factory=get_sql_database_blob_auditing_policies_operations) as g:
 
-        g.show_command('show', 'get')
-        g.generic_update_command('update',
-                                 custom_func_name='db_audit_policy_update')
+        g.custom_show_command('show', 'db_audit_policy_show')
+        g.custom_command('update', 'db_audit_policy_update')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/commands.py:264]: May I know the reason changing generic_update_command to custom_command? With this change, you will lose some parameters like --set, --add, --remove.

File: src/azure-cli/azure/cli/command_modules/sql/commands.py [hunk 12]
```diff
@@ -272,9 +272,9 @@ def load_command_table(self, _):
                             server_blob_auditing_policies_operations,
                             client_factory=get_sql_server_blob_auditing_policies_operations) as g:
 
-        g.show_command('show', 'get')
-        g.generic_update_command('update',
-                                 custom_func_name='server_audit_policy_update')
+        g.custom_show_command('show', 'server_audit_policy_show')
+        g.custom_command('update', 'server_audit_policy_update', supports_no_wait=True)
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/commands.py:276]: same generic update issue

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 13]
```diff
@@ -43,6 +43,11 @@
     ServerPublicNetworkAccess
 )
 
+from azure.mgmt.monitor.models import (RetentionPolicy, LogSettings)
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py:46]: For monitor related operation usage, please consider get_models s below and use ResourceType.MGMT_MONITOR
https://github.com/Azure/azure-cli/blob/45fae7733dc832e77fa7d4dee6542d802ea2b160/src/azure-cli/azure/cli/command_modules/network/custom.py#L1880

**Review by Juliehzl — APPROVED** [review]
LGTM in general

**Pull request merged**